pkg/querier/tripperware/metadata/get_series.go

```diff
@@ -0,0 +1,289 @@
+package metadata
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/gogo/status"
+	jsoniter "github.com/json-iterator/go"
+	"github.com/opentracing/opentracing-go"
+	otlog "github.com/opentracing/opentracing-go/log"
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/timestamp"
+	"github.com/weaveworks/common/httpgrpc"
+
+	"github.com/cortexproject/cortex/pkg/querier/tripperware"
+	"github.com/cortexproject/cortex/pkg/querier/tripperware/queryrange"
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/cortexproject/cortex/pkg/util/spanlogger"
+)
+
+var (
+	GetSeriesCodec tripperware.Codec = newGetSeriesCodec()
+	json                             = jsoniter.Config{
+		EscapeHTML:             false, // No HTML in our responses.
+		SortMapKeys:            true,
+		ValidateJsonRawMessage: true,
+	}.Froze()
+)
+
+type PrometheusGetSeriesResponseExtended struct {
+	PrometheusGetSeriesResponse
+	Data []map[string]string `json:"data"`
+}
+
+type PrometheusGetSeriesRequest struct {
+	tripperware.Request
+	Path     string
+	Start    int64
+	End      int64
+	Matchers []string
+	Headers  http.Header
+}
+
+// WithStartEnd clones the current `PrometheusRequest` with a new `start` and `end` timestamp.
+func (r *PrometheusGetSeriesRequest) WithStartEnd(start int64, end int64) tripperware.Request {
+	new := *r
+	new.Start = start
+	new.End = end
+	return &new
+}
+
+func (r *PrometheusGetSeriesRequest) GetStart() int64 {
+	return r.Start
+}
+
+func (r *PrometheusGetSeriesRequest) GetEnd() int64 {
+	return r.End
+}
+
+// GetStep returns always 0 for Series.
+func (r *PrometheusGetSeriesRequest) GetStep() int64 {
+	return 0
+}
+
+// LogToSpan writes information about this request to an OpenTracing span
+func (r *PrometheusGetSeriesRequest) LogToSpan(sp opentracing.Span) {
+	sp.LogFields(
+		otlog.String("match[]", fmt.Sprintf("%v", r.Matchers)),
+		otlog.String("start", timestamp.Time(r.GetStart()).String()),
+		otlog.String("end", timestamp.Time(r.GetEnd()).String()),
+	)
+}
+
+type getSeriesCodec struct {
+	tripperware.Codec
+	now func() time.Time
+}
+
+func newGetSeriesCodec() getSeriesCodec {
+	return getSeriesCodec{now: time.Now}
+}
+
+func (resp *PrometheusGetSeriesResponseExtended) HTTPHeaders() map[string][]string {
+	if resp != nil && resp.GetHeaders() != nil {
+		r := map[string][]string{}
+		for _, header := range resp.GetHeaders() {
+			if header != nil {
+				r[header.Name] = header.Values
+			}
+		}
+
+		return r
+	}
+	return nil
+}
+
+func (c getSeriesCodec) DecodeRequest(_ context.Context, r *http.Request, forwardHeaders []string) (tripperware.Request, error) {
+	result := PrometheusGetSeriesRequest{Headers: map[string][]string{}}
+	var err error
+	result.Start, err = parseTimeParam(r, "start", c.now().Add(-1*time.Hour).Unix())
```

Comment:
It is better to make this lookback duration configurable

Reply:
Let me first explain why I had to do this. The defaults for the official Prometheus API is 

```
minTime = time.Unix(math.MinInt64/1000+62135596801, 0).UTC()
maxTime = time.Unix(math.MaxInt64/1000-62135596801, 999999999).UTC()
```

But if we do this here, we'll have thousands of 24h splits. Instead I decided to set sane defaults for these.


I'll make it configurable.